# Описание проекта
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.



**Описание данных:**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**
Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**Навыки и инструменты:**
- pandas
- sqlalchemy

## Подключение к базе

In [1]:
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
con=engine.connect()

In [3]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books'''
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

## Исследование таблиц

### Таблица books

In [4]:
query = '''
            SELECT * 
            FROM books
            LIMIT (5)
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query = '''
            SELECT count(*) 
            FROM books
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

count
0   1000

В таблице `books` 1000 строк. Содержит id книги, id автора, название книги, кол-во страниц, дата публикации и id издателя.

### Таблица authors

In [6]:
query = '''
            SELECT * 
            FROM authors
            LIMIT(5)
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
query = '''
            SELECT count(*) 
            FROM authors
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    636

В таблице authors 636 строк, содержит id автора и данные автора.

### Таблица publishers

In [8]:
query = '''
            SELECT * 
            FROM publishers
            LIMIT(5)
        '''
pd.io.sql.read_sql(sql=text(query), con = con)


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
query = '''
            SELECT count(*) 
            FROM publishers
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    340

Датасет содержит 340 строк, также информацию об издателе и его id номера.

### Таблица ratings

In [10]:
query = '''
            SELECT * 
            FROM ratings
            LIMIT(5)
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
query = '''
            SELECT count(*) 
            FROM ratings
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

count
0   6456

В датасете 6456 строк, содержит информацию о id оценки,id книги,имя пользователя, оставившего оценку и оценку книги.

### Таблица reviews

In [12]:
query = '''
            SELECT * 
            FROM reviews
            LIMIT(5)
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
query = '''
            SELECT count(*) 
            FROM reviews
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

count
0   2793

Датафрейм содержит 2793 строк, информацию об id обзора, id книги, имя пользователя, написавшего обзор и текст обзора.

## Cколько книг вышло после 1 января 2000 года

In [14]:
query = '''
            SELECT COUNT(book_id) count_of_books
            FROM books 
            WHERE publication_date > '01-01-2000'
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

count_of_books
0             819

После 1 января 2000 года вышло 819 книг.

## Количество обзоров и среднюю оценку

Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [15]:
pd.set_option('max_colwidth', 200)
pd.set_option('display.width', 300)
query = '''
            SELECT b.title, rat.avg_rating, rev.review_count
            FROM books b
            LEFT JOIN 
            (SELECT book_id, AVG(rating) avg_rating 
            FROM ratings 
            GROUP BY book_id) rat ON b.book_id=rat.book_id
            LEFT JOIN 
            (SELECT book_id, 
            COUNT(*) review_count 
            FROM reviews 
            GROUP BY book_id) rev ON b.book_id=rev.book_id
            ORDER BY avg_rating DESC
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

title  avg_rating  review_count
0                                                                                      Pop Goes the Weasel (Alex Cross  #5)        5.00           2.0
1    The Ghost Map: The Story of London's Most Terrifying Epidemic—and How It Changed Science  Cities  and the Modern World        5.00           2.0
2                                                                      In the Hand of the Goddess (Song of the Lioness  #2)        5.00           2.0
3                                                                                                  Tai-Pan (Asian Saga  #2)        5.00           2.0
4                                                       How to Be a Domestic Goddess: Baking and the Art of Comfort Cooking        5.00           1.0
..                                                                                                                      ...         ...           ...
995                                                          The World Is Flat: A Brief History of the Twenty-first Century        2.25           3.0
996                                                                                                                   Junky        2.00           2.0
997                                                                                       His Excellency: George Washington        2.00           2.0
998                                                                                                           Drowning Ruth        2.00           3.0
999                                                                                                    Harvesting the Heart        1.50           2.0

[1000 rows x 3 columns]

По таблице видно, что наименьший рейтинг у книг 1.5, высший бал - 5.0. Также в среднем книги имеют по несколько отзывов.
Наиболее высокую оценку получили книги: 
- Pop Goes the Weasel (Alex Cross #5)
- The Ghost Map: The Story of London's Most Terrifying Epidemic—and How It Changed Science Cities and the Modern World
- In the Hand of the Goddess (Song of the Lioness #2)	
- Tai-Pan (Asian Saga #2)

## Определение издательства, которое выпустило наибольшее число книг толще 50 страниц

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [16]:
query = '''
            SELECT b.publisher_id, p.publisher, 
            COUNT(*) count_books
            FROM books b
            LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY b.publisher_id, p.publisher
            ORDER BY COUNT(*) DESC
            LIMIT 1
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id      publisher  count_books
0           212  Penguin Books           42

Издательство `Penguin Books` выпустило 42 книги толще 50 страниц.

## Определение автора с самой высокой средней оценкой книг

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [17]:
query = '''
            SELECT a.author, 
            AVG(ratings)
            FROM (
            SELECT book_id, AVG(rating) ratings
            FROM ratings
            GROUP BY book_id
            HAVING COUNT(*) > 50
            ) r
            JOIN (SELECT book_id, author_id FROM books) b ON r.book_id=b.book_id
            JOIN (SELECT author_id, author FROM authors) a ON b.author_id=a.author_id
            GROUP BY a.author
            ORDER BY avg(ratings) DESC
            LIMIT 5
            
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

author       avg
0         J.K. Rowling/Mary GrandPré  4.283844
1  Markus Zusak/Cao Xuân Việt Khương  4.264151
2                     J.R.R. Tolkien  4.258446
3                  Louisa May Alcott  4.192308
4                       Rick Riordan  4.080645

Наиболее высокие оценки у авторов J.K. Rowling/Mary GrandPré.

## Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок.

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [18]:
query = '''
            SELECT AVG(count_reviews)
            FROM (
            SELECT username, COUNT(*) count_reviews
            FROM reviews
            WHERE username IN
            (
            SELECT username--, COUNT(*)
            FROM ratings
            GROUP BY username
            HAVING COUNT(*) > 48
            )
            GROUP BY username) rev
        

        '''
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

Среднее кол-во пользователей, которые поставили больше 48 оценок - 24 пользователя.

## Доп.задания

Выведите таблицу, которая будет содержать по году публикации:

- количество издательств,

- выпущенных книг и

- сколько всего тысяч страниц было в изданных книгах

(отобразить только те года, в которых издано более 30 книг)



In [19]:
query = '''
            SELECT 
            CAST(EXTRACT(YEAR FROM publication_date) AS int) AS year,
            COUNT (DISTINCT(publisher_id)) count_publisher,
            COUNT(book_id) count_books,
            SUM(num_pages)
            FROM books 
            GROUP BY year
            HAVING COUNT(book_id) > 30;
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

year  count_publisher  count_books    sum
0  1999               26           41  15763
1  2000               35           38  13328
2  2001               41           60  21758
3  2002               62           94  38597
4  2003               65          105  41423
5  2004               88          124  46779
6  2005               89          139  55967
7  2006              109          184  68302
8  2007               38           50  18258

Выведите в одной таблице два числа — среднюю оценку тех книг, на которые написало отзывов более 3 человек и отдельно среднюю оценку остальных книг, сделайте выводы какой рейтинг больше


In [20]:
query = '''
SELECT
    (SELECT AVG(rating) 
     FROM ratings 
     WHERE book_id IN (
         SELECT book_id 
         FROM reviews 
         GROUP BY book_id 
         HAVING COUNT(book_id) > 3
     )
    ) AS avg_rating_3_reviews,
    (SELECT AVG(rating) 
     FROM ratings 
     WHERE book_id IN (
         SELECT book_id 
         FROM reviews 
         GROUP BY book_id 
         HAVING COUNT(book_id) < 3
     )
    ) AS avg_rating_other;
    
'''
pd.io.sql.read_sql(sql=text(query), con = con)

avg_rating_3_reviews  avg_rating_other
0              3.949505          3.906279

Больше средний рейтинг у книг с отзывами больше 3.

Выведите топ пять пользователей по суммарному показателю написанных ревью и поставленных оценок

In [21]:
query = '''
            SELECT rat.username,
            (ratings + reviews) AS count
            FROM (
            SELECT username, Count(rating) ratings
            FROM ratings
            GROUP BY username
            ) rat
            LEFT JOIN
            (SELECT username, Count(reviews) reviews
            FROM reviews
            GROUP BY username) rev ON rat.username=rev.username
            ORDER BY count DESC
            LIMIT 5;
            
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

username  count
0      martinadam     83
1     sfitzgerald     83
2       richard89     81
3         susan85     78
4  jennifermiller     78

## Вывод:

Мы проанализировали табдицы и получили следующие выводы:

- После 1 января 2000 года вышло 819 книг.
- Наименьший рейтинг у книг 1.5, высший бал - 5.0. Также в среднем книги имеют по несколько отзывов.
- Из всех издательств, больше всего книг толще 50 страниц выпустило издательство Penguin Books.
- Самые высокие оценки у авторов J.K. Rowling/Mary GrandPré.
- Среднее кол-во пользователей, которые поставили больше 48 оценок - 24 пользователя.